interfaces/prompting/constraints.go

```diff
@@ -30,62 +31,128 @@ import (
 	"github.com/snapcore/snapd/strutil"
 )
 
-// Constraints hold information about the applicability of a rule to particular
-// paths or permissions. A request matches the constraints if the requested path
-// is matched by the path pattern (according to bash's globstar matching) and
-// the requested permissions are contained in the constraints' permissions.
+// Constraints hold information about the applicability of a new rule to
+// particular paths or permissions. A request will be matched by the constraints
+// if the requested path is matched by the path pattern (according to bash's
+// globstar matching) and one or more requested permissions are denied in the
+// permission map, or all of the requested permissions are allowed in the map.
+// When snapd creates a new rule, it converts Constraints to RuleConstraints.
 type Constraints struct {
-	PathPattern *patterns.PathPattern `json:"path-pattern,omitempty"`
-	Permissions []string              `json:"permissions,omitempty"`
+	PathPattern *patterns.PathPattern `json:"path-pattern"`
+	Permissions PermissionMap         `json:"permissions"`
 }
 
-// ValidateForInterface returns nil if the constraints are valid for the given
-// interface, otherwise returns an error.
-func (c *Constraints) ValidateForInterface(iface string) error {
+// ToRuleConstraints validates the receiving Constraints and converts it to
+// RuleConstraints. If the constraints are not valid with respect to the given
+// interface, returns an error.
+func (c *Constraints) ToRuleConstraints(iface string, currTime time.Time) (*RuleConstraints, error) {
+	if c.PathPattern == nil {
+		return nil, prompting_errors.NewInvalidPathPatternError("", "no path pattern")
+	}
+	rulePermissions, err := c.Permissions.ToRulePermissionMap(iface, currTime)
```

Comment:
ToRulePermissionMap seems to be used only here, I think we want to export only methods that are used more than once/can be used elsewhere

Reply:
Yes, this method can be unexported, thanks! I'll check if I can find any more which should not be exported.